data/src/main/java/com/linkedin/data/schema/compatibility/AnnotationCompatibilityChecker.java

```diff
@@ -0,0 +1,133 @@
+   */
+  static List<AnnotationCompatibilityResult> getCompatibilityResult(Map<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext,
+      Map<String, Object>>> olderResolvedPropertiesMap, Map<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext,
+      Map<String, Object>>> newerResolvedPropertiesMap, List<SchemaAnnotationHandler> handlers)
+  {
+    List<AnnotationCompatibilityResult> results = new ArrayList<>();
+    for (SchemaAnnotationHandler handler: handlers)
+    {
+      olderResolvedPropertiesMap.entrySet().stream().forEach(e ->
+          {
+            String annotationNamespace = handler.getAnnotationNamespace();
+            PathSpec pathSpec = e.getKey();
+            if (newerResolvedPropertiesMap.containsKey(pathSpec))
+            {
+              Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>> olderCheckContextAndResolvedProperty
+                  = olderResolvedPropertiesMap.get(pathSpec);
+              Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>> newerCheckContextAndResolvedProperty
+                  = newerResolvedPropertiesMap.get(pathSpec);
+
+              Map<String, Object> olderResolvedProperties = olderCheckContextAndResolvedProperty.getValue();
+              Map<String, Object> newerResolvedProperties = newerCheckContextAndResolvedProperty.getValue();
+
+              if (olderResolvedProperties.containsKey(annotationNamespace) || newerResolvedProperties.containsKey(annotationNamespace))
+              {
+                AnnotationCompatibilityResult result = handler.annotationCompatibilityCheck(olderResolvedProperties,
+                    newerResolvedProperties, olderCheckContextAndResolvedProperty.getKey(), newerCheckContextAndResolvedProperty.getKey());
+                results.add(result);
+              }
+            }
+          });
+    }
+    return results;
+  }
+
+  private static Map<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>> getNodeToResolvedProperties(
+      SchemaAnnotationProcessor.SchemaAnnotationProcessResult result)
+  {
+    AnnotationCheckResolvedPropertiesVisitor visitor = new AnnotationCheckResolvedPropertiesVisitor();
+    DataSchemaRichContextTraverser traverser = new DataSchemaRichContextTraverser(visitor);
+    traverser.traverse(result.getResultSchema());
+    return visitor.getNodeToResolvedPropertiesMap();
+  }
+
+  private static SchemaAnnotationProcessor.SchemaAnnotationProcessResult processSchemaAnnotation(DataSchema dataSchema,
+      List<SchemaAnnotationHandler> handlers)
+  {
+    SchemaAnnotationProcessor.SchemaAnnotationProcessResult result =
+        SchemaAnnotationProcessor.process(handlers, dataSchema, new SchemaAnnotationProcessor.AnnotationProcessOption());
+    // If any of the nameDataSchema failed to be processed, throw exception
+    if (result.hasError())
+    {
+      String schemaName = ((NamedDataSchema) dataSchema).getFullName();
```

Comment:
Add log before throwing exception

Reply:
log and throw exception is anti-pattern, right ? https://stackoverflow.com/questions/6639963/why-is-log-and-throw-considered-an-anti-pattern

Reply:
Added log